private network configuration guide (#72)
sourcegraph/docs: docs/admin/config/private_network.mdx

# Private network configuration

> When deploying self-hosted Sourcegraph instances in private networks with specific compliance and policy requirements, additional configuration may be required to ensure all networking features are functioning correctly. The reasons for applying the following configuration options depend on the specific functionality of the Sourcegraph service and the unique network and infrastructure requirements of the organization. The following Sourcegraph services typically initiate outbound connections to external services:

Author: good call out, added the following for clarity:
> A "private network" refers to a secure network environment that is segregated from the public internet, designed to facilitate internal communications and operations within an organization. This network setup restricts external access, thereby enhancing security and control over data flow by limiting exposure to external threats and unauthorized access.

> - **sourcegraph-frontend**: The frontend service is responsible for connecting to external [auth providers](../auth), sending [telemetry data](../pings), testing code host connections, and connecting to [externally hosted](../external_services) Sourcegraph services

Reviewer: "testing code host connections" 
I didn't realize frontend did this.  Does it connect just to verify the connectivity to the provider?  or is this for GitHub app only?

> - **sourcegraph-frontend**: The frontend service is responsible for connecting to external [auth providers](../auth), sending [telemetry data](../pings), testing code host connections, and connecting to [externally hosted](../external_services) Sourcegraph services

Author: I'll confirm this with engineering, but I observed this to be the case while working with a customer to troubleshoot this [issue](https://github.com/sourcegraph/sourcegraph/pull/49372).

CheckConnection interface:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph@012bc5bcf1d6b575ebcfc782545df85a8d6c6e5d/-/blob/internal/repos/sources.go?L98-101

- **gitserver**: Executes git commands against externally hosted [code hosts](../external_service)

> - **repo-updater**: Communicates with [code hosts](../external_service) APIs to coordinate the repository synchronize

Reviewer: "the repository synchronization"?

### Kubernetes Helm
Add the proxy environment variables to your Sourcegraph Helm chart [override file](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml):

```yaml
frontend|gitserver|repo-updater:
  env:
  - name: HTTP_PROXY 
    value: http://proxy.example.com:8080
  - name: HTTPS_PROXY
    value: http://proxy.example.com:8080
  - name: NO_PROXY
    value:  "*.cluster.local,blobstore,codeinsights-db,codeintel-db,sourcegraph-frontend-internal,sourcegraph-frontend,github-proxy,gitserver,grafana,indexed-search-indexer,indexed-search,jaeger-query,pgsql,precise-code-intel-worker,prometheus,redis-cache,redis-store,repo-updater,searcher,symbols,syntect-server,worker-executors,worker" 
```

<Callout type="warning">Failure to configure `NO_PROXY` correctly can cause proxy configuration to interfere with local networking between internal Sourcegraph services</Callout>

## Using private CA root certificates

> Some organizations maintain a private Certificate Authority (CA) for issuing certificates within their private network. When Sourcegraph connects to TLS encrypted service using a self-signed certificate that it does not trust, it will throw the `x509: certificate signed by unknown authority` error message in logs.

Reviewer: minor nitpick: ", it will throw the ..."  ===> ", you will see `x509: certificate signed by unknown authority` error messages in the logs"

# Private network configuration

> A "private network" refers to a secure network environment that is segregated from the public internet, designed to facilitate internal communications and operations within an organization. This network setup restricts external access, thereby enhancing security and control over data flow by limiting exposure to external threats and unauthorized access.

Reviewer 2: ```suggestion
A **private network** refers to a secure network environment segregated from the public internet, designed to facilitate internal communications and operations within an organization. This network setup restricts external access, enhancing security and control over data flow by limiting exposure to external threats and unauthorized access.
```

> When deploying self-hosted Sourcegraph instances in private networks with specific compliance and policy requirements, additional configuration may be required to ensure all networking features are functioning correctly. The reasons for applying the following configuration options depend on the specific functionality of the Sourcegraph service and the unique network and infrastructure requirements of the organization. The following is a list of Sourcegraph services and how and when each initiates outbound connections to external services:

Reviewer 2: ```suggestion
When deploying self-hosted Sourcegraph instances in private networks with specific compliance and policy requirements, additional configuration may be required to ensure all networking features function correctly. The reasons for applying the following configuration options depend on the specific functionality of the Sourcegraph service and the unique network and infrastructure requirements of the organization. 

The following is a list of Sourcegraph services and how and when each initiates outbound connections to external services:
```

### Kubernetes Helm
Add the proxy environment variables to your Sourcegraph Helm chart [override file](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml):

```yaml
frontend|gitserver|repo-updater:
  env:
  - name: HTTP_PROXY 
    value: http://proxy.example.com:8080
  - name: HTTPS_PROXY
    value: http://proxy.example.com:8080
  - name: NO_PROXY
    value:  "*.cluster.local,blobstore,codeinsights-db,codeintel-db,sourcegraph-frontend-internal,sourcegraph-frontend,github-proxy,gitserver,grafana,indexed-search-indexer,indexed-search,jaeger-query,pgsql,precise-code-intel-worker,prometheus,redis-cache,redis-store,repo-updater,searcher,symbols,syntect-server,worker-executors,worker" 
```

<Callout type="warning">Failure to configure `NO_PROXY` correctly can cause proxy configuration to interfere with local networking between internal Sourcegraph services</Callout>

## Using private CA root certificates

> Some organizations maintain a private Certificate Authority (CA) for issuing certificates within their private network. When Sourcegraph connects to TLS encrypted service using a self-signed certificate that it does not trust, you will observe a `x509: certificate signed by unknown authority` error message in logs.

Reviewer 2: ```suggestion
Some organizations maintain a private Certificate Authority (CA) for issuing certificates within their private network. When Sourcegraph connects to TLS encrypted service using a self-signed certificate that it does not trust, you will observe an `x509: certificate signed by unknown authority` error message in logs.
```
